src/core/PlaneWaveBasis.jl

```diff
@@ -0,0 +1,216 @@
+
+## Examples
+```julia-repl
+julia> b = PlaneWaveBasis(TODO)
+```
+
+## Arguments and struct fields
+- `lattice`: Real-space lattice vectors in columns
+- `recip_lattice`: Reciprocal lattice vectors in columns
+- `grid_size`:     Size of the rectangular Fourier grid used as the
+                   density basis ``B_ρ``. In each dimension `idim` the
+                   range of wave vectors (in integer coordinates) extends from
+                   `-ceil(Int, (grid_size[idim]-1) / 2)` up to
+                   `floor(Int, (grid_size[idim]-1) / 2)`. No optimisation is done
+                   on the size of the grid with respect to performing FFTs.
+- `Ecut`:          Kinetic energy cutoff in Hartree
+- `kpoints`:       List of ``k``-Points in fractional coordinats
+- `kweights`:      List of corresponding weights for the Brillouin-zone integration.
+- `kbasis`:        List of wave vectors (in integer coordinates) for each ``k``-Point,
+                   chosen such that ``|G + k|^2/2 \leq Ecut`` where ``G`` is the
+                   wave vector. The defines the wave function basis ``B_{Ψ,k}`` for
+                   each ``k``-Point.
+- `FFT` and `iFFT`:   Containers for the forward and backward FFT plans on ``B_ρ``.
+"""
+struct PlaneWaveBasis{T <: Real}
+    lattice::SMatrix{3, 3, T}
+    recip_lattice::SMatrix{3, 3, T}
+    unit_cell_volume::T
+    recip_cell_volume::T
+
+    grid_size::SVector{3, Int}
+    Ecut::T
+    kpoints::Vector{SVector{3, T}}
+    kweights::Vector{T}
+    kbasis::Vector{Vector{SVector{3, Int}}}
+
+    FFT
+    iFFT
+end
+
+function PlaneWaveBasis(lattice::AbstractMatrix{T}, grid_size,
+                        Ecut::Number, kpoints, kweights) where {T <: Real}
+    lattice = SMatrix{3, 3, T}(lattice)
+    recip_lattice = 2π * inv(lattice')
+
+    # Plan a FFT, spending some time on finding an optimal algorithm
+    # for the machine on which the computation runs
+    grid_size = SVector{3, Int}(grid_size)
+    tmp = Array{Complex{T}}(undef, grid_size...)
+    fft_plan = plan_fft!(tmp, flags=FFTW.MEASURE)
+    ifft_plan = plan_ifft!(tmp, flags=FFTW.MEASURE)
+
+    dummy_kpoints = Vector{SVector{3, T}}(undef, length(kpoints))
+    dummy_kweights = Vector{T}(undef, length(kweights))
+    dummy_kbasis = Vector{Vector{SVector{3, Int}}}(undef, length(kpoints))
+    pw = PlaneWaveBasis{T}(lattice, recip_lattice, det(lattice), det(recip_lattice),
+                           grid_size, Ecut, dummy_kpoints, dummy_kweights, dummy_kbasis,
+                           fft_plan, ifft_plan)
+    substitute_kpoints!(pw, kpoints, kweights)
+end
+
+"""
+Replace the kpoints of an existing Plane-wave basis and make
+the internal data structures consistent.
+"""
+function substitute_kpoints!(pw::PlaneWaveBasis{T}, kpoints, kweights) where T
+    if length(kpoints) != length(kweights)
+        error("Lengths of kpoints and length of kweights need to agree")
+    end
+    kweights = Vector{T}(kweights / sum(kweights))  # Normalise kweights
```

Comment:
If the "user" gives an unnormalized list of weights, they might have their reasons, and it seems dangerous to silently ignore that. If you want to avoid funny business here you can have an `@assert sum(kweights) ≈ 1`

Reply:
Hmm ... I see the point. I make an assertion.